src/lib.rs

```diff
@@ -216,19 +213,17 @@ fn encode_dag_cbor_from_pyobject<'py, W: Write>(
         }
 
         Ok(())
-    } else if obj.is_instance_of::<PyList>() {
-        let seq = obj.downcast::<PySequence>().unwrap();
-        let len = obj.len()?;
+    } else if let Ok(l) = obj.downcast::<PyList>() {
+        let len = l.len();
 
         encode::write_u64(w, MajorKind::Array, len as u64)?;
 
         for i in 0..len {
-            encode_dag_cbor_from_pyobject(py, &seq.get_item(i)?, w)?;
+            encode_dag_cbor_from_pyobject(py, &l.get_item(i)?, w)?;
         }
 
         Ok(())
-    } else if obj.is_instance_of::<PyDict>() {
-        let map = obj.downcast::<PyMapping>().unwrap();
+    } else if let Ok(map) = obj.downcast::<PyMapping>() {
```

Comment:
Downcast to `Mapping` can run a lot of python code, this is probably the perf issue. Maybe the PyO3 documentation can be improved.

Reply:
you are real Jedi! fixed and updated. gained perf boost as expected! so useful with a lot of floats (canada fixture)